File: probes/published-spec/adapter.py
"""One JSON document in; public installed Python spec observations out."""
import hashlib
import importlib.metadata
import json
import platform
import sys

import agent_topology.spec as spec

raw = sys.stdin.buffer.read()
result = {
    "identity": {
        "version": importlib.metadata.version("agent-topology-spec"),
        "runtime": platform.python_version(),
        "imported_path": spec.__file__,
        "executable": sys.executable,
    },
    "input_sha256": hashlib.sha256(raw).hexdigest(),
}
stage = "parse"
try:
    document = json.loads(raw)
    stage = "validation"
    errors = spec.validate_document(document)
    result["validation"] = {"accepted": not errors, "errors": errors}
    if errors:
        result["status"] = "rejected"
    else:
        stage = "canonical"
        result["canonical"] = spec.canonical_json(document)
        stage = "hash"
        result["structureHash"] = spec.compute_structure_hash(document)
        result["status"] = "accepted"
except Exception as error:
    result.update(status="error", error={"stage": stage, "type": type(error).__name__, "message": str(error)})
print(json.dumps(result, ensure_ascii=False))
sys.exit(0 if result["status"] == "accepted" else 2 if result["status"] == "rejected" else 3)

File: probes/published-spec/adapter.mjs
// Copied into each isolated installation so bare imports resolve there.
import { readFileSync } from 'node:fs';
import { createHash } from 'node:crypto';
import * as spec from '@agent-topology/spec';

const raw = readFileSync(0);
const result = {
  identity: {
    version: JSON.parse(readFileSync(new URL('./node_modules/@agent-topology/spec/package.json', import.meta.url))).version,
    runtime: process.version,
    imported_path: import.meta.resolve('@agent-topology/spec'),
    executable: process.execPath,
  },
  input_sha256: createHash('sha256').update(raw).digest('hex'),
};
let stage = 'parse';
try {
  const document = JSON.parse(raw.toString('utf8'));
  stage = 'validation';
  const validation = spec.validateDocument(document);
  result.validation = { accepted: validation.valid, errors: validation.errors };
  if (!validation.valid) result.status = 'rejected';
  else {
    stage = 'canonical';
    result.canonical = spec.canonicalStringify(document);
    stage = 'hash';
    result.structureHash = spec.computeStructureHash(document);
    result.status = 'accepted';
  }
} catch (error) {
  result.status = 'error';
  result.error = { stage, type: error.name, message: error.message };
}
console.log(JSON.stringify(result));
process.exitCode = result.status === 'accepted' ? 0 : result.status === 'rejected' ? 2 : 3;
